src/main/java/org/kiwiproject/registry/consul/server/ConsulRegistryService.java

```diff
@@ -0,0 +1,220 @@
+                .maxAttempts(MAX_UNREGISTER_ATTEMPTS)
+                .retryDelayTime(UNREGISTER_RETRY_DELAY)
+                .retryDelayUnit(UNREGISTER_RETRY_DELAY_UNIT)
+                .build();
+    }
+
+    @Override
+    public ServiceInstance createCandidateFrom(ServiceInfo serviceInfo) {
+        return ServiceInstance.fromServiceInfo(serviceInfo)
+                .withStatus(ServiceInstance.Status.UP);
+    }
+
+    @Override
+    public ServiceInstance register(ServiceInstance serviceToRegister) {
+        checkState(!isRegistered(), "Cannot register. Already managing a registered instance");
+
+        var registration = fromServiceInstance(serviceToRegister);
+
+        var service = registerRetryer.tryGetObject(() -> {
+            consul.agentClient().register(registration);
+            return serviceToRegister.withInstanceId(registration.getId());
+        });
+
+        if (service.isEmpty()) {
+            LOG.error("Registration failed for service {} with id {}. See logs above", serviceToRegister.getServiceName(), registration.getId());
+            var errMsg = format("Unable to register service %s, id %s with Consul after %s attempts",
+                    serviceToRegister.getServiceName(), registration.getId(), MAX_REGISTRATION_ATTEMPTS);
+            throw new RegistrationException(errMsg);
+        }
+
+        registeredService.set(service.get());
+        return registeredService.get().toBuilder().build();
+    }
+
+    @Override
+    public ServiceInstance updateStatus(ServiceInstance.Status newStatus) {
+        LOG.info("Ignoring status update as Consul handles it's own status through it's health check process");
+
+        if (isRegistered()) {
+            return registeredService.get().toBuilder().build();
+        }
+
+        return null;
+    }
+
+    @Override
+    public void unregister() {
+        checkState(isRegistered(), "Cannot unregister since registration was never called");
+
+        var serviceName = registeredService.get().getServiceName();
+        var instanceId = registeredService.get().getInstanceId();
+
+        LOG.info("Unregistering service {} with id {}", serviceName, instanceId);
+
+        var result = unregisterRetryer.tryGetObject(() -> {
+            consul.agentClient().deregister(registeredService.get().getInstanceId());
+            return instanceId;
+        });
+
+        Optionals.ifPresentOrElseThrow(result,
+                id -> {
+                    LOG.info("Service with name {} and id {} has been unregistered successfully", serviceName, instanceId);
+                    registeredService.set(null);
+                },
+                () -> {
+                    var msg = format("Error un-registering service {}, id {}", serviceName, instanceId);
+                    LOG.error(msg);
+                    return new RegistrationException(msg);
+                });
+    }
+
+    private boolean isRegistered() {
+        return nonNull(registeredService.get());
```

Comment:
I have now seen `registeredService.get()` at least a half-dozen times. Maybe pull that expression out into a small private helper method, e.g.

```java
private ServiceInstance getServiceInstance() {
    return registeredService.get();
}
```

Reply:
Done